Fixes #9016 - improved fact loading performance

lzap opened the pull request:
This change improves fact loading performance by order of magnitude on
SQLite mostly thanks to explicit transaction. An extra optimization was
done for the most important "update" method call which is mostly used
during fact uploads.

Updating of 500 existing facts is faster from 14.8 seconds to 2.7
seconds with this patch.

The patch utilizes find_each method with works in batches rather than creating
instance for each individual model object. More reading at:

http://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each

lzap committed:
Fixes #9016 - improved fact loading performance

This change improves fact loading performance by order of magnitude on
SQLite mostly thanks to explicit transaction. An extra optimization was
done for the most important "update" method call which is mostly used
during fact uploads.

Updating of 500 existing facts is faster from 14.8 seconds to 2.7
seconds with this patch.

lzap commented:
It looks like explicit transaction has huge impact on SQLite while on PostgreSQL there is no impact. The other part of the patch (`find_each` method) have some impact on SQLite while it is less noticeable on PostgreSQL. In both cases, memory consumption should be lower.

Average response time of updating 1000 facts per call on production deployment with the patch is:
- develop branch: 2200 ms
- with this patch without explicit transaction: 2100 ms
- with this patch incl. explicit transaction: 2100 ms

This has almost zero effect on new discovered hosts (they all add new facts and update none), but existing discover hosts which refresh facts should be also improved.

If you find more bottlenecks in our code, let me know. I considered using https://github.com/zdennis/activerecord-import for `add_new_facts` but since we need to get IDs per individual FactName, it is not good fit.

Another option would be to use `create!([{...}, {...}])` method, but I tried it with Rails 4.2 and it does seem to use two separate SQL INSERTs instead of one INSERT ... VALUES call.

lzap commented:
@ares @ohadlevy I suggest to carry on with the discussion here.

The motivation is that updating 1000 facts takes 2 seconds on production bare-metal server and servers tend to report hundreds (e.g. 300 is pretty much standard number) of facts. And this is quite important Foreman call I suppose (update facts via PuppetMaster).

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -91,16 +93,18 @@ def create_fact_name(fact_names, name, fact_value)
+    db_facts_as_relation.find_each do |record|
+      new_value = facts[record.name]
+      if record.value != new_value
+        record.update_attribute(:value, new_value)
```
any reason not to use `update_column` here?

ohadlevy replied:
what is the performance difference? this would skip any call backs or hooking on fact value change.

tbrisker replied:
`update_column` just sends an SQL `UPDATE`, if we don't need callbacks it should be much faster then updating the active record and saving it.

ohadlevy replied:
hence why i'm asking for the difference, I'm not sure we actually have any
callbacks (besides people who might use foreman hooks) but avoiding it does
not let us handle events in the future (e.g. if we want a notification of
certain fact changes etc).

lzap replied:
> hence why i'm asking for the difference, I'm not sure we actually have any
> callbacks (besides people who might use foreman hooks) but avoiding it does
> not let us handle events in the future (e.g. if we want a notification of
> certain fact changes etc).

To be honest, I actually do like to skip callbacks rather than to
introduce performance degradation.

Anyway, do you see any another bottlenecks in the importing code?

## 

Later,
 Lukas #lzap Zapletal

ohadlevy replied:
i guess its the multiple inserts/updates, in theory we could have done it in one - this is pretty old but probably still relevant: https://www.coffeepowered.net/2009/01/23/mass-inserting-data-in-rails-without-killing-your-performance/

if you skip callbacks, you can in theory look on other options.

tbrisker replied:
I think the only thing more that I think can be optimized is if we can somehow update all the facts in one query, but I'm not sure it's possible in this case.
Note that by skipping the callbacks we don't get the updated_at field updated, but it can be updated in one query using the `update_all` method.

lzap replied:
Yeah, I don't like writing own SQL, so I think we can leave it for now. I think Rails will support this at some point and we can take advantage of it in the future.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -111,7 +115,11 @@ def normalize(facts)
+
   def db_facts
-    @db_facts ||= host.fact_values.eager_load(:fact_name).where("fact_names.type = '#{fact_name_class}'").index_by(&:name)
+    @db_facts ||= db_facts_as_relation.index_by(&:name)
```
The only use AFAICT for this function is above in `add_new_facts` - where we only pick out the names anyways. 
So I think it could be rewritten as `db_facts_as_relation.pluck('fact_names.name')` and this method and instance variable can be discarded. 
only other consideration is if other plugins depend on this function, not sure.

lzap replied:
But this method is only called once, I don't think this is worth changing to pluck, we need to create hash from it anyway (`index_by` creates a hash).

tbrisker replied:
Why do we need the hash? we only use the hash keys in the `add_new_facts` method, is there anywhere we use the hash values after your change?

lzap replied:
Youre right, foreman_salt does use it as well, but I can fix that.

lzap commented:
So to sum it up, I will change the patch to use `update_column` and `delete_all` to skip callbacks and that's it, if there are no objections.

lzap commented:
I just found that `audited` is disabled for FactValues, therefore we can optimize it for direct access and skip callbacks/validations. I will implement `updated_at` mass update as @tbrisker recommended, good idea.

Now, obviously fact updates is too heavy for Audited gem, how about to create a simple Observer API around this so users can hook into it for their own audits via simple plugins. A reference implementation could be sending changes into standard logger.

lzap commented:
Amended, created instrumentation points for plugins to dump fact changes out of foreman (e.g. into external database or files). Reference implementation uses `LogSubscriber` to send the information to newly created `audit` logger similarly to LDAP.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -37,9 +37,11 @@ def initialize(host, facts = {})
+    ActiveRecord::Base.transaction do
+      delete_removed_facts
+      add_new_facts
+      update_facts
```
I think you should switch the order to first update then add, otherwise you are going to check if the new facts were changed (which they obviously have not) during the update.

lzap replied:
I can do that.

tbrisker commented on config/application.rb:
```diff
@@ -170,7 +170,8 @@ class Application < Rails::Application
       :permissions => {:enabled => false},
-      :sql => {:enabled => false}
+      :sql => {:enabled => false},
+      :audit => {:enabled => true}
```
This is a confusing name for a logger that deals with facts, not audits

lzap replied:
Well, I am hoping to create a dedicated logger for all audit logs that are too noisy for audited gem. Fact importer is just the first thing here. For the same reason I named the initializer as `audit_*.rb`.

tbrisker replied:
Nevermind, I figured this out. still trying to think if we could simplify this function somehow as it feels a bit over complicated to me.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -81,26 +82,35 @@ def add_new_facts
+      end
+    end
+    # set updated_at field explicitly
+    FactValue.where(:id => updated.map(&:id)).update_all(:updated_at => Time.now)
```
after thinking about this a bit more, we're already updating each of these rows above, so why not use:

``` ruby
record.update_columns(:value => new_value, :updated_at => Time.now)
```

lzap replied:
Yeah, I will change it.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -55,17 +57,16 @@ def fact_name_class
+    deleted_count = FactValue.joins(:fact_name).where(:host => host, 'fact_names.type' => fact_name_class).where.not('fact_names.name' => facts.keys).delete_all
+    @counters[:deleted] = deleted_count
+    ActiveSupport::Notifications.instrument "fact_importer_deleted.foreman", :host_id => host.id, :host_name => host.name, :facts => facts, :count => deleted_count do
+      logger.debug("Merging facts for '#{host}': deleted #{deleted_count} facts")
```
why do we log here the same thing that is logged by the listener but with a different level?

lzap replied:
Yeah, I will change that and put the actual code into it, this way I can also print execution time of it for performance reasons (instrumentation code calculate it anyway I will only print it).

lzap replied:
This is how it looks now with duration:

```
2016-07-27T10:46:19 [audit] [I] [host1] deleted 0 facts (4.52ms)
2016-07-27T10:46:19 [audit] [I] [host1] updated 0 facts {} (3.08ms)
2016-07-27T10:46:19 [audit] [I] [host1] added 1 facts {"_timestamp"=>"234"} (14.05ms)
2016-07-27T10:46:19 [audit] [I] [host1] deleted 0 facts (0.78ms)
2016-07-27T10:46:19 [audit] [I] [host1] updated 0 facts {} (0.98ms)
2016-07-27T10:46:19 [audit] [I] [host1] added 1 facts {"_timestamp"=>"345"} (4.39ms)
2016-07-27T10:46:20 [audit] [I] [host2] deleted 0 facts (1.4ms)
2016-07-27T10:46:20 [audit] [I] [host2] updated 0 facts {} (1.57ms)
2016-07-27T10:46:20 [audit] [I] [host2] added 1 facts {"foo"=>"bar"} (4.23ms)
2016-07-27T10:46:20 [audit] [I] [host3] deleted 1 facts (0.8ms)
2016-07-27T10:46:20 [audit] [I] [host3] updated 1 facts {"ipaddress"=>"10.0.19.5"} (1.61ms)
2016-07-27T10:46:20 [audit] [I] [host3] added 1 facts {"uptime"=>"1 picosecond"} (4.17ms)
2016-07-27T10:46:20 [audit] [I] [host4] deleted 1 facts (0.96ms)
2016-07-27T10:46:20 [audit] [I] [host4] updated 0 facts {} (1.3ms)
2016-07-27T10:46:20 [audit] [I] [host4] added 2 facts {"vda_size"=>"4242", "structured"=>"{\"key\"=>\"value\"}"} (7.12ms)
2016-07-27T10:46:20 [audit] [I] [host5] deleted 1 facts (0.79ms)
2016-07-27T10:46:20 [audit] [I] [host5] updated 0 facts {} (1.23ms)
2016-07-27T10:46:20 [audit] [I] [host5] added 0 facts {} (0.51ms)
2016-07-27T10:46:20 [audit] [I] [host6] deleted 1 facts (0.88ms)
2016-07-27T10:46:20 [audit] [I] [host6] updated 0 facts {} (1.28ms)
2016-07-27T10:46:20 [audit] [I] [host6] added 0 facts {} (0.56ms)
2016-07-27T10:46:20 [audit] [I] [host7] deleted 0 facts (0.79ms)
2016-07-27T10:46:20 [audit] [I] [host7] updated 1 facts {"kernelversion"=>"3.8.11"} (1.68ms)
2016-07-27T10:46:20 [audit] [I] [host7] added 0 facts {} (0.52ms)
2016-07-27T10:46:20 [audit] [I] [host8] deleted 1 facts (0.86ms)
2016-07-27T10:46:20 [audit] [I] [host8] updated 1 facts {"ipaddress"=>"10.0.19.5"} (1.63ms)
2016-07-27T10:46:20 [audit] [I] [host8] added 1 facts {"uptime"=>"1 picosecond"} (4.2ms)
```

tbrisker commented on config/initializers/audit_instrumentation.rb:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    def filter_facts(event, collection)
+      facts = event.payload[:facts].select {|k, v| event.payload[collection].include?(k)}.inspect
```
I'm guessing this should be `@facts ||= event....`?

lzap replied:
No, it's single call with unique input (event).

lzap replied:
The variable is useless, will remove.

lzap commented:
Amended all your comments, added duration, removed debug logs, broke complex method into two.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -112,6 +115,6 @@ def normalize(facts)
 
   def db_facts
-    @db_facts ||= host.fact_values.eager_load(:fact_name).where("fact_names.type = '#{fact_name_class}'").index_by(&:name)
+    host.fact_values.eager_load(:fact_name).where("fact_names.type = '#{fact_name_class}'")
```
Is there any reason not to use `.joins` (for an inner join) instead of `.eager_load` (which generates a left outer join)?

lzap replied:
No, it works just fine. Lemme change that.

tbrisker commented on app/services/fact_importer.rb:
```diff
@@ -55,52 +57,53 @@ def fact_name_class
+      # if the host does not exists yet, we don't have an host_id to use the fact_values table.
+      if facts_to_create.present?
+        # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+        fact_names = fact_name_class.where(:type => fact_name_class).index_by(&:name)
```
Would it make sense to also replace the `index_by(&:name)` here with `pluck(:name)`? I don't see any use for this hash other then for deciding which names need to be created.

lzap commented:
Have you deleted some comments from here? I am confused, I see some via email but not here.

tbrisker commented:
@lzap yeah, sorry for that. I tried to reply to myself with a "nevermind" that somehow got attached to the wrong comment. Ended up just removing those two comments as they were basically "why do this?" "oh nevermind, figured it out", so nothing substantial in them.

tbrisker commented:
@lzap did you try benchmarking with the latest version? I don't know how much overhead the logging adds, we should make sure it isn't even slower then adding audits.

lzap commented:
Yeah, no noticeable change in times, still 2100 ms. I rebased.

lzap commented:
I created a micro-benchmark to measure improvement in this patch. It's not big. Develop:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                          6.896  (± 0.0%) i/s -    104.000  in  15.140540s
puppet fact importer - adding unique facts (size 1)
                          6.735  (±14.8%) i/s -    102.000  in  15.236680s
puppet fact importer - adding facts with same values (size 1)
                          6.753  (±14.8%) i/s -    102.000  in  15.235439s
puppet fact importer - updating facts with same names (size 1)
                          6.782  (±14.7%) i/s -    102.000  in  15.163636s
puppet fact importer - adding and deleting facts (size 1)
                          6.670  (± 0.0%) i/s -    100.000  in  15.036979s
puppet fact importer - adding facts (size 200)
                          2.159  (± 0.0%) i/s -     33.000  in  15.324107s
puppet fact importer - adding unique facts (size 200)
                          1.114  (± 0.0%) i/s -     17.000  in  15.372189s
puppet fact importer - adding facts with same values (size 200)
                          1.009  (± 0.0%) i/s -     16.000  in  15.975508s
puppet fact importer - updating facts with same names (size 200)
                          1.526  (± 0.0%) i/s -     23.000  in  15.161395s
puppet fact importer - adding and deleting facts (size 200)
                          1.463  (± 0.0%) i/s -     22.000  in  15.150607s
puppet fact importer - adding facts (size 1000)
                          0.497  (± 0.0%) i/s -      8.000  in  16.159521s
puppet fact importer - adding unique facts (size 1000)
                          0.253  (± 0.0%) i/s -      4.000  in  15.817914s
puppet fact importer - adding facts with same values (size 1000)
                          0.239  (± 0.0%) i/s -      4.000  in  16.764849s
puppet fact importer - updating facts with same names (size 1000)
                          0.444  (± 0.0%) i/s -      7.000  in  15.835309s
puppet fact importer - adding and deleting facts (size 1000)
                          0.426  (± 0.0%) i/s -      7.000  in  16.500930s
```

This PR:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                          6.689  (±14.9%) i/s -    100.000  in  15.039273s
puppet fact importer - adding unique facts (size 1)
                          6.538  (±15.3%) i/s -     98.000  in  15.086233s
puppet fact importer - adding facts with same values (size 1)
                          6.250  (±16.0%) i/s -     94.000  in  15.142077s
puppet fact importer - updating facts with same names (size 1)
                          6.330  (±15.8%) i/s -     96.000  in  15.286524s
puppet fact importer - adding and deleting facts (size 1)
                          6.265  (± 0.0%) i/s -     94.000  in  15.070470s
puppet fact importer - adding facts (size 200)
                          2.327  (± 0.0%) i/s -     35.000  in  15.065722s
puppet fact importer - adding unique facts (size 200)
                          1.231  (± 0.0%) i/s -     19.000  in  15.542241s
puppet fact importer - adding facts with same values (size 200)
                          1.109  (± 0.0%) i/s -     17.000  in  15.423205s
puppet fact importer - updating facts with same names (size 200)
                          1.627  (± 0.0%) i/s -     25.000  in  15.480967s
puppet fact importer - adding and deleting facts (size 200)
                          1.397  (± 0.0%) i/s -     21.000  in  15.160986s
puppet fact importer - adding facts (size 1000)
                          0.597  (± 0.0%) i/s -      9.000  in  15.090487s
puppet fact importer - adding unique facts (size 1000)
                          0.301  (± 0.0%) i/s -      5.000  in  16.619438s
puppet fact importer - adding facts with same values (size 1000)
                          0.285  (± 0.0%) i/s -      5.000  in  17.546525s
puppet fact importer - updating facts with same names (size 1000)
                          0.514  (± 0.0%) i/s -      8.000  in  15.624756s
puppet fact importer - adding and deleting facts (size 1000)
                          0.397  (± 0.0%) i/s -      6.000  in  15.115661s
```

I will do some conclusion tomorrow, when I want to take a look on one extra patch discussed in the Issue ticket.

lzap commented:
Analysis: As you can see, the current patch has little impact on time, numbers (these are transactions per second) are just a little bit better. I can push numbers little bit further by applying one other change which was inspired by the patch from Jon McKenzie (see the associated issue):

``` diff
diff --git a/app/services/fact_importer.rb b/app/services/fact_importer.rb
index 7cb21d5..ad2f462 100644
--- a/app/services/fact_importer.rb
+++ b/app/services/fact_importer.rb
@@ -63,9 +63,9 @@ class FactImporter
     end
   end

-  def add_new_fact(fact_names, name)
+  def add_new_fact(name)
     method = host.new_record? ? :build : :create!
-    fact_name = create_fact_name(fact_names, name, facts[name])
+    fact_name = fact_name_class.where(:name => name, :type => fact_name_class).first_or_create!
     host.fact_values.send(method, :value => facts[name], :fact_name => fact_name)
   rescue => e
     logger.error("Fact #{name} could not be imported because of #{e.message}")
@@ -77,19 +77,13 @@ class FactImporter
       facts_to_create = facts.keys - db_facts.pluck('fact_names.name')
       # if the host does not exists yet, we don't have an host_id to use the fact_values table.
       if facts_to_create.present?
-        # :type is needed because custom facts usually inherits from FactName so they would be included in the list
-        fact_names = fact_name_class.where(:type => fact_name_class).index_by(&:name)
-        facts_to_create.each { |f| add_new_fact(fact_names, f) }
+        facts_to_create.each { |f| add_new_fact(f) }
       end
       payload[:added] = facts_to_create
       payload[:count] = @counters[:added] = facts_to_create.size
     end
   end

-  def create_fact_name(fact_names, name, fact_value)
-    fact_names[name] ||= fact_name_class.create!(:name => name)
-  end
-
   def update_facts
     ActiveSupport::Notifications.instrument "fact_importer_updated.foreman", :host_id => host.id, :host_name => host.name, :facts => facts do |payload|
       time = Time.now
```

This significantly gives better results for fact imports with small amount of facts changed:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                         22.558  (± 4.4%) i/s -    338.000  in  15.044736s
puppet fact importer - adding unique facts (size 1)
                         21.728  (± 4.6%) i/s -    326.000  in  15.062249s
puppet fact importer - adding facts with same values (size 1)
                         21.442  (± 4.7%) i/s -    322.000  in  15.077930s
puppet fact importer - updating facts with same names (size 1)
                         22.133  (± 4.5%) i/s -    332.000  in  15.054449s
puppet fact importer - adding and deleting facts (size 1)
                         20.313  (± 4.9%) i/s -    304.000  in  15.033666s
puppet fact importer - adding facts (size 200)
                          2.242  (± 0.0%) i/s -     34.000  in  15.183026s
puppet fact importer - adding unique facts (size 200)
                          1.307  (± 0.0%) i/s -     20.000  in  15.322042s
puppet fact importer - adding facts with same values (size 200)
                          1.258  (± 0.0%) i/s -     19.000  in  15.155820s
puppet fact importer - updating facts with same names (size 200)
                          2.222  (± 0.0%) i/s -     34.000  in  15.320483s
puppet fact importer - adding and deleting facts (size 200)
                          1.726  (± 0.0%) i/s -     26.000  in  15.080587s
puppet fact importer - adding facts (size 1000)
                          0.472  (± 0.0%) i/s -      8.000  in  16.970766s
puppet fact importer - adding unique facts (size 1000)
                          0.269  (± 0.0%) i/s -      5.000  in  18.587890s
puppet fact importer - adding facts with same values (size 1000)
                          0.267  (± 0.0%) i/s -      5.000  in  18.771698s
puppet fact importer - updating facts with same names (size 1000)
                          0.472  (± 0.0%) i/s -      8.000  in  16.991459s
puppet fact importer - adding and deleting facts (size 1000)
                          0.356  (± 0.0%) i/s -      6.000  in  16.892674s
```

I will publish the bencharking script shortly, you will note that I haven't pre-created any facts and values in the database. Testing with empty database was a starting point which does not show full potential of the patch. Also, what haven't been measured by this is memory usage, which should be significantly better with the patch.

lzap commented:
I am pushing the above change and the benchmark as well. I also added 10,000 fact names prior test runs, so results are more appropriate. Without the patch:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                          3.017  (± 0.0%) i/s -      4.000  in   1.333497s
puppet fact importer - adding unique facts (size 1)
                          2.542  (± 0.0%) i/s -      3.000  in   1.216058s
puppet fact importer - adding facts with same values (size 1)
                          2.894  (± 0.0%) i/s -      4.000  in   1.402985s
puppet fact importer - updating facts with same names (size 1)
                          3.012  (± 0.0%) i/s -      4.000  in   1.335795s
puppet fact importer - adding and deleting facts (size 1)
                          2.523  (± 0.0%) i/s -      3.000  in   1.221507s
puppet fact importer - adding facts (size 200)
                          1.570  (± 0.0%) i/s -      2.000  in   1.273972s
puppet fact importer - adding unique facts (size 200)
                          0.948  (± 0.0%) i/s -      1.000  in   1.055142s
puppet fact importer - adding facts with same values (size 200)
                          0.981  (± 0.0%) i/s -      1.000  in   1.019575s
puppet fact importer - updating facts with same names (size 200)
                          1.543  (± 0.0%) i/s -      2.000  in   1.296523s
puppet fact importer - adding and deleting facts (size 200)
                          1.497  (± 0.0%) i/s -      2.000  in   1.336000s
puppet fact importer - adding facts (size 1000)
                          0.511  (± 0.0%) i/s -      1.000  in   1.955520s
puppet fact importer - adding unique facts (size 1000)
                          0.249  (± 0.0%) i/s -      1.000  in   4.013732s
puppet fact importer - adding facts with same values (size 1000)
                          0.245  (± 0.0%) i/s -      1.000  in   4.089364s
puppet fact importer - updating facts with same names (size 1000)
                          0.484  (± 0.0%) i/s -      1.000  in   2.064054s
puppet fact importer - adding and deleting facts (size 1000)
                          0.453  (± 0.0%) i/s -      1.000  in   2.206700s
Memory stats
Total objects allocated: 64763901
Total heap pages allocated: 4408
```

With the current patch:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                         24.630  (± 4.1%) i/s -     26.000  in   1.057426s
puppet fact importer - adding unique facts (size 1)
                         23.617  (± 4.2%) i/s -     24.000  in   1.017307s
puppet fact importer - adding facts with same values (size 1)
                         23.429  (± 4.3%) i/s -     24.000  in   1.025411s
puppet fact importer - updating facts with same names (size 1)
                         24.237  (±12.4%) i/s -     24.000  in   1.005980s
puppet fact importer - adding and deleting facts (size 1)
                         22.601  (± 4.4%) i/s -     24.000  in   1.063056s
puppet fact importer - adding facts (size 200)
                          2.366  (± 0.0%) i/s -      3.000  in   1.269185s
puppet fact importer - adding unique facts (size 200)
                          1.392  (± 0.0%) i/s -      2.000  in   1.436310s
puppet fact importer - adding facts with same values (size 200)
                          1.403  (± 0.0%) i/s -      2.000  in   1.425321s
puppet fact importer - updating facts with same names (size 200)
                          2.327  (± 0.0%) i/s -      3.000  in   1.289317s
puppet fact importer - adding and deleting facts (size 200)
                          1.838  (± 0.0%) i/s -      2.000  in   1.087981s
puppet fact importer - adding facts (size 1000)
                          0.506  (± 0.0%) i/s -      1.000  in   1.974653s
puppet fact importer - adding unique facts (size 1000)
                          0.291  (± 0.0%) i/s -      1.000  in   3.439825s
puppet fact importer - adding facts with same values (size 1000)
                          0.292  (± 0.0%) i/s -      1.000  in   3.423827s
puppet fact importer - updating facts with same names (size 1000)
                          0.506  (± 0.0%) i/s -      1.000  in   1.976361s
puppet fact importer - adding and deleting facts (size 1000)
                          0.388  (± 0.0%) i/s -      1.000  in   2.574575s
Memory stats
Total objects allocated: 60674797
Total heap pages allocated: 4411
```

That's about six million ruby objects less allocated than before with the same or better throughput than before (significantly better for smallish requests).

lzap commented:
Rebased, this is ready for review.

lzap commented on test/benchmark/benchmark_helper.rb:
```diff
@@ -0,0 +1,29 @@
+require 'factory_girl_rails'
+
+ unless Rails.env.production? and ActiveRecord::Base.connection.class == ActiveRecord::ConnectionAdapters::SQLite3Adapter and Rails.configuration.database_configuration[Rails.env]['database'] == ":memory:"
+  require 'pry'; binding.pry
```
> > -  require 'pry'; binding.pry
> 
> Remove debugger entry point `binding.pry`.<br>Use 2 (not 1) spaces for indentation.<br>Do not use semicolons to terminate expressions.

LOL You!

## 

Later,
 Lukas #lzap Zapletal

lzap commented on test/benchmark/benchmark_helper.rb:
```diff
@@ -0,0 +1,29 @@
+  timeout: 1000
+EOS
+  exit 1
+end
```
Fixed, updated.

This is how you actually run the benchmark:

```
FOREMAN_NO_DEPRECATIONS=1 DISABLE_SPRING=true RAILS_ENV=production be ruby -Itest test/benchmark/puppet_fact_importer_benchmark.rb
```

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
Fixed the other Hound issue (nice one actually). Rebased!

lzap commented on test/benchmark/puppet_fact_importer_benchmark.rb:
```diff
@@ -0,0 +1,51 @@
+require "benchmark/benchmark_helper"
```
What on Earth is this?

lzap replied:
Relevant patch is https://github.com/theforeman/foreman/pull/3845

lzap commented:
Rebased due to frozen literals.

tbrisker commented:
@lzap Looks like tests are failing.

tbrisker commented on test/benchmark/benchmark_helper.rb:
```diff
@@ -0,0 +1,29 @@
+  timeout: 1000
+EOS
+  exit 1
+end
```
Why do we run tests in production rather then test environment?

lzap replied:
Well it's a benchmark rather than test.

dLobatog replied:
`FOREMAN_NO_DEPRECATIONS` does what exactly? I don't see any instance of it by grepping the dir.

tbrisker replied:
still, why should it run in production and require changing the production env just to run it? and for that matter, why should it only be limited to in-memory sqlite? wouldn't we want to benchmark on other DBs as well?

dLobatog commented on test/benchmark/benchmark_helper.rb:
```diff
@@ -0,0 +1,28 @@
+require 'benchmark/ips'
```
Add to bundler.d/development.rb , this is not part of the standard lib

lzap replied:
Sure

dLobatog commented on test/benchmark/benchmark_helper.rb:
```diff
@@ -0,0 +1,28 @@
+require File.expand_path('../../../config/environment', __FILE__)
+require 'factory_girl_rails'
+
+unless Rails.env.production? && ActiveRecord::Base.connection.class == ActiveRecord::ConnectionAdapters::SQLite3Adapter && Rails.configuration.database_configuration[Rails.env]['database'] == ":memory:"
```
You can't do that check and expect the error message to be printed out if it's not true, as that adapter won't be loaded.
e.g: my production.yml contains `adapter: mysql`, then I get an error `uninitialized constant ActiveRecord::ConnectionAdapters::SQLite3Adapter (NameError)` when trying to run this code instead of the message

lzap replied:
Yeah, I will only check for production and migration configuration value.

dLobatog commented on test/benchmark/puppet_fact_importer_benchmark.rb:
```diff
@@ -0,0 +1,51 @@
+
+foreman_benchmark do
+  Benchmark.ips do |x|
+    #x.config(:time => 15, :warmup => 2)
```
?

lzap replied:
Leftover, actually this is the correct setting and the other one (1/1) was just for testing the benchmark itself. I see you executed the tests with 1/1, can you try again with 15/2? That should give better results. Also, are you testing with MySQL?

dLobatog replied:
Yup I'm using mysql to test it

dLobatog commented on config/initializers/audit_instrumentation.rb:
```diff
@@ -0,0 +1,32 @@
+# Audit logging from fact importer events sent via ActiveSupport::Notifications
```
More docs explaining how `.instrument` stuff would go into this logger would be helpful, I don't expect everyone to know about it right away

lzap replied:
It is not generic, so I don't expects other classes to use it. We already have similar initializer. It's explained here: http://api.rubyonrails.org/classes/ActiveSupport/Notifications.html I am not sure what you expect me to add.

dLobatog commented:
I've run stats on 10000 - looks good but not only that, I like that we can now tweak the importer and figure out what to optimize, thanks to the stats.

This is the **new**   importer:

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                         21.695  (±13.8%) i/s -     22.000  in   1.043781s
puppet fact importer - adding unique facts (size 1)
                         19.114  (±20.9%) i/s -     20.000  in   1.087804s
puppet fact importer - adding facts with same values (size 1)
                         19.858  (±15.1%) i/s -     20.000  in   1.027246s
puppet fact importer - updating facts with same names (size 1)
                         22.011  (± 9.1%) i/s -     22.000  in   1.008998s
puppet fact importer - adding and deleting facts (size 1)
                         19.962  (±15.0%) i/s -     20.000  in   1.039862s
puppet fact importer - adding facts (size 200)
                          1.951  (± 0.0%) i/s -      2.000  in   1.028645s
puppet fact importer - adding unique facts (size 200)
                          1.058  (± 0.0%) i/s -      2.000  in   1.894748s
puppet fact importer - adding facts with same values (size 200)
                          1.055  (± 0.0%) i/s -      2.000  in   1.895970s
puppet fact importer - updating facts with same names (size 200)
                          1.953  (± 0.0%) i/s -      2.000  in   1.024046s
puppet fact importer - adding and deleting facts (size 200)
                          1.633  (± 0.0%) i/s -      2.000  in   1.226501s
puppet fact importer - adding facts (size 1000)
                          0.210  (± 0.0%) i/s -      1.000  in   4.755075s
puppet fact importer - adding unique facts (size 1000)
                          0.218  (± 0.0%) i/s -      1.000  in   4.578555s
puppet fact importer - adding facts with same values (size 1000)
                          0.251  (± 0.0%) i/s -      1.000  in   3.992006s
puppet fact importer - updating facts with same names (size 1000)
                          0.448  (± 0.0%) i/s -      1.000  in   2.230890s
puppet fact importer - adding and deleting facts (size 1000)
                          0.297  (± 0.0%) i/s -      1.000  in   3.365227s
puppet fact importer - adding facts (size 10000)
                          0.034  (± 0.0%) i/s -      1.000  in  29.463571s
puppet fact importer - adding unique facts (size 10000)
                          0.015  (± 0.0%) i/s -      1.000  in  68.579230s
puppet fact importer - adding facts with same values (size 10000)
                          0.020  (± 0.0%) i/s -      1.000  in  49.958916s
puppet fact importer - updating facts with same names (size 10000)
                          0.035  (± 0.0%) i/s -      1.000  in  28.761106s
puppet fact importer - adding and deleting facts (size 10000)
                          0.020  (± 0.0%) i/s -      1.000  in  49.505848s
Memory stats
Total objects allocated: 282581021
Total heap pages allocated: 7884
```

vs the **old** importer

```
Calculating -------------------------------------
puppet fact importer - adding facts (size 1)
                          1.156  (± 0.0%) i/s -      2.000  in   1.763244s
puppet fact importer - adding unique facts (size 1)
                          1.128  (± 0.0%) i/s -      2.000  in   1.781051s
puppet fact importer - adding facts with same values (size 1)
                          0.348  (± 0.0%) i/s -      1.000  in   2.876629s
puppet fact importer - updating facts with same names (size 1)
                          0.694  (± 0.0%) i/s -      2.000  in   4.003225s
puppet fact importer - adding and deleting facts (size 1)
                          1.041  (± 0.0%) i/s -      2.000  in   1.921579s
puppet fact importer - adding facts (size 200)
                          0.296  (± 0.0%) i/s -      1.000  in   3.382335s
puppet fact importer - adding unique facts (size 200)
                          0.542  (± 0.0%) i/s -      1.000  in   1.844590s
puppet fact importer - adding facts with same values (size 200)
                          0.536  (± 0.0%) i/s -      1.000  in   1.864258s
puppet fact importer - updating facts with same names (size 200)
                          0.646  (± 0.0%) i/s -      1.000  in   1.547216s
puppet fact importer - adding and deleting facts (size 200)
                          0.603  (± 0.0%) i/s -      1.000  in   1.657352s
puppet fact importer - adding facts (size 1000)
                          0.269  (± 0.0%) i/s -      1.000  in   3.722216s
puppet fact importer - adding unique facts (size 1000)
                          0.133  (± 0.0%) i/s -      1.000  in   7.505896s
puppet fact importer - adding facts with same values (size 1000)
                          0.157  (± 0.0%) i/s -      1.000  in   6.387504s
puppet fact importer - updating facts with same names (size 1000)
                          0.170  (± 0.0%) i/s -      1.000  in   5.887417s
puppet fact importer - adding and deleting facts (size 1000)
                          0.160  (± 0.0%) i/s -      1.000  in   6.242237s
puppet fact importer - adding facts (size 10000)
                          0.023  (± 0.0%) i/s -      1.000  in  43.927541s
puppet fact importer - adding unique facts (size 10000)
                          0.013  (± 0.0%) i/s -      1.000  in  77.266427s
puppet fact importer - adding facts with same values (size 10000)
                          0.017  (± 0.0%) i/s -      1.000  in  60.040679s
puppet fact importer - updating facts with same names (size 10000)
                          0.033  (± 0.0%) i/s -      1.000  in  30.136244s
puppet fact importer - adding and deleting facts (size 10000)
                          0.032  (± 0.0%) i/s -      1.000  in  31.606148s
Memory stats
Total objects allocated: 262577993
Total heap pages allocated: 12407
```

Great stuff @lzap ! It doesn't seem to change much on the 10k case though?

lzap commented:
Ok fixed structured fact importer. Crossing fingers.

lzap commented:
> Great stuff @lzap ! It doesn't seem to change much on the 10k case though?

Yes, as I've said the main purpose and improvement is the  memory consumption (as you can see in the output).

lzap commented:
Pushed https://gist.github.com/80b212ea3a7e980b00d48ec5db6c35ab

lzap commented on app/services/structured_fact_importer.rb:
```diff
@@ -19,6 +19,28 @@ def normalize_recurse(memo, facts, prefix = '')
+    fact_names[name] ||= fact_name_class.create!(:name => name)
+  end
+
   def create_fact_name(fact_names, name, fact_value)
```
Oh these methods are not necessary anymore. I will remove that in the next push, they are unused.

dLobatog replied:
:+1:

lzap reviewed (commented):
Damn accidentally started a review...

dLobatog commented on app/services/structured_fact_importer.rb:
```diff
@@ -19,6 +19,28 @@ def normalize_recurse(memo, facts, prefix = '')
   end
 
+  def find_or_create_fact_name(name, value)
+    if name.include?(FactName::SEPARATOR)
```
What is this doing? It tries to find who is the parent fact or create it when the factname has a separator?  
Extracting to a separate method with a name that explains what it does and/or comments would help here IMO

lzap replied:
This is just refactor of the current create_fact_name (see the second implementation below, the one I added was just a miscopy).

I don't agree with extracting this method, this is a compact recursive method. Codeclimate is simply wrong here. Let me push and show you how it looks like.

I am also adding more benchmarks to test the structured importer to see the difference.

dLobatog reviewed (commented):
@lzap Two small changes, extracting the 'find or create nested fact' and removing the duplicate `create_fact_name` and I think it'd be ready, thanks!

lzap commented:
Pushing https://gist.github.com/1ab7dc4bec315a2b5e508488f6d798eb

Can you perform the benchmark on MySQL now with and without the patch?

lzap commented on test/benchmark/puppet_fact_importer_benchmark.rb:
```diff
@@ -0,0 +1,71 @@
+end
+
+foreman_benchmark do
+  [StructuredFactImporter]
```
Opps leftover, one more push.

dLobatog commented on app/services/structured_fact_importer.rb:
```diff
@@ -19,19 +19,21 @@ def normalize_recurse(memo, facts, prefix = '')
   end
 
-  def create_fact_name(fact_names, name, fact_value)
+  def find_or_create_fact_name(name, value)
```
Can you make value optional `value = nil`? It seems like you're calling it within the method with `,nil)` when you need to create a fact.

Also can you add a comment to explain what exactly it's doing?

lzap commented:
Benchmark of the latest version on my i7 T430s, develop branch:

```
PuppetFactImporter: adding facts (size 1)
                          3.285  (± 0.0%) i/s -     49.000  in  15.010622s
PuppetFactImporter: adding unique facts (size 1)
                          3.259  (± 0.0%) i/s -     49.000  in  15.145479s
PuppetFactImporter: adding facts with same values (size 1)
                          3.266  (± 0.0%) i/s -     49.000  in  15.083348s
PuppetFactImporter: updating facts with same names (size 1)
                          3.281  (± 0.0%) i/s -     49.000  in  15.033419s
PuppetFactImporter: adding and deleting facts (size 1)
                          3.280  (± 0.0%) i/s -     49.000  in  15.017206s
StructuredFactImporter: adding facts (size 1)
                          3.265  (± 0.0%) i/s -     49.000  in  15.090598s
StructuredFactImporter: adding unique facts (size 1)
                          3.283  (± 0.0%) i/s -     49.000  in  15.016064s
PuppetFactImporter: adding facts (size 200)
                          1.562  (± 0.0%) i/s -     24.000  in  15.416030s
PuppetFactImporter: adding unique facts (size 200)
                          1.557  (± 0.0%) i/s -     24.000  in  15.475979s
PuppetFactImporter: adding facts with same values (size 200)
                          1.551  (± 0.0%) i/s -     24.000  in  15.529554s
PuppetFactImporter: updating facts with same names (size 200)
                          1.553  (± 0.0%) i/s -     24.000  in  15.509975s
PuppetFactImporter: adding and deleting facts (size 200)
                          1.480  (± 0.0%) i/s -     23.000  in  15.648368s
StructuredFactImporter: adding facts (size 200)
                          1.528  (± 0.0%) i/s -     23.000  in  15.094801s
StructuredFactImporter: adding unique facts (size 200)
                          1.532  (± 0.0%) i/s -     23.000  in  15.064097s
PuppetFactImporter: adding facts (size 1000)
                          0.484  (± 0.0%) i/s -      8.000  in  16.566708s
PuppetFactImporter: adding unique facts (size 1000)
                          0.477  (± 0.0%) i/s -      8.000  in  16.803251s
PuppetFactImporter: adding facts with same values (size 1000)
                          0.476  (± 0.0%) i/s -      8.000  in  16.818147s
PuppetFactImporter: updating facts with same names (size 1000)
                          0.474  (± 0.0%) i/s -      8.000  in  16.911639s
PuppetFactImporter: adding and deleting facts (size 1000)
                          0.441  (± 0.0%) i/s -      7.000  in  15.926300s
StructuredFactImporter: adding facts (size 1000)
                          0.415  (± 0.0%) i/s -      7.000  in  17.348360s
StructuredFactImporter: adding unique facts (size 1000)
                          0.431  (± 0.0%) i/s -      7.000  in  16.309081s
Memory stats
Total objects allocated: 323198482
Total heap pages allocated: 4408
```

This PR:

```
PuppetFactImporter: adding facts (size 1)
                         24.018  (± 4.2%) i/s -    360.000  in  15.037099s
PuppetFactImporter: adding unique facts (size 1)
                         23.972  (± 4.2%) i/s -    360.000  in  15.060334s
PuppetFactImporter: adding facts with same values (size 1)
                         23.783  (± 4.2%) i/s -    356.000  in  15.047316s
PuppetFactImporter: updating facts with same names (size 1)
                         23.652  (± 4.2%) i/s -    354.000  in  15.028711s
PuppetFactImporter: adding and deleting facts (size 1)
                         22.045  (± 4.5%) i/s -    330.000  in  15.004667s
StructuredFactImporter: adding facts (size 1)
                         23.574  (± 4.2%) i/s -    354.000  in  15.061067s
StructuredFactImporter: adding unique facts (size 1)
                         23.592  (± 4.2%) i/s -    354.000  in  15.055990s
PuppetFactImporter: adding facts (size 200)
                          1.398  (± 0.0%) i/s -     21.000  in  15.026510s
PuppetFactImporter: adding unique facts (size 200)
                          1.381  (± 0.0%) i/s -     21.000  in  15.218668s
PuppetFactImporter: adding facts with same values (size 200)
                          1.384  (± 0.0%) i/s -     21.000  in  15.174550s
PuppetFactImporter: updating facts with same names (size 200)
                          1.382  (± 0.0%) i/s -     21.000  in  15.207399s
PuppetFactImporter: adding and deleting facts (size 200)
                          1.186  (± 0.0%) i/s -     18.000  in  15.178716s
StructuredFactImporter: adding facts (size 200)
                          1.395  (± 0.0%) i/s -     21.000  in  15.054691s
StructuredFactImporter: adding unique facts (size 200)
                          1.374  (± 0.0%) i/s -     21.000  in  15.304721s
PuppetFactImporter: adding facts (size 1000)
                          0.291  (± 0.0%) i/s -      5.000  in  17.177695s
PuppetFactImporter: adding unique facts (size 1000)
                          0.292  (± 0.0%) i/s -      5.000  in  17.148820s
PuppetFactImporter: adding facts with same values (size 1000)
                          0.292  (± 0.0%) i/s -      5.000  in  17.125310s
PuppetFactImporter: updating facts with same names (size 1000)
                          0.292  (± 0.0%) i/s -      5.000  in  17.128048s
PuppetFactImporter: adding and deleting facts (size 1000)
                          0.249  (± 0.0%) i/s -      4.000  in  16.069672s
StructuredFactImporter: adding facts (size 1000)
                          0.291  (± 0.0%) i/s -      5.000  in  17.154443s
StructuredFactImporter: adding unique facts (size 1000)
                          0.292  (± 0.0%) i/s -      5.000  in  17.112130s
Memory stats
Total objects allocated: 376390260
Total heap pages allocated: 4403
```

As you can see, the patch improves memory allocation (less objects allocated), as a drawback it is slightly slower with huge facts sets (500+ facts in one query). There is a huge performance gain for small queries (only one fact), but this is corner case that is not much relevant.

I am not fully confident with the benchmark I wrote, I think I will create another to be close to real world scenarios.

lzap commented:
I missed an issue in the audit instrumentor - when an error occurs, payload was not set properly. Also logging facts into debug now.

And I rewrote the benchmark from the scratch, it now tests with 200 and 500 number of fact in each import which is more relevant and it tests with 0 and 50 unique fact names and also with 0 and 25 structured facts (of length 10 in format `fX::fX::fX::fX::fX::fX::fX::fX::fX::fX` where X is a random number). This is repeated for normal importer and structured importer.

https://gist.github.com/5488bf9b9384bb001a6ef80a52b5f1da

lzap commented:
I have rebased the patch on top of (https://github.com/theforeman/foreman/pull/3852), which essentially removes the code added by the other patch because we are changing approach in this. This patch seems to be faster and consume less memory. It almost doubled the performance against develop.

Develop:

```
PuppetFactImporter (200) - 0 UN 0 SN
                          0.989  (± 0.0%) i/s -     10.000  in  10.282023s
PuppetFactImporter (200) - 0 UN 25 SN
                          0.930  (± 0.0%) i/s -     10.000  in  10.785003s
PuppetFactImporter (200) - 50 UN 0 SN
                          0.892  (± 0.0%) i/s -     10.000  in  11.226758s
PuppetFactImporter (200) - 50 UN 25 SN
                          0.856  (± 0.0%) i/s -      9.000  in  10.519111s
PuppetFactImporter (500) - 0 UN 0 SN
                          0.558  (± 0.0%) i/s -      6.000  in  10.836860s
PuppetFactImporter (500) - 0 UN 25 SN
                          0.546  (± 0.0%) i/s -      6.000  in  11.004525s
PuppetFactImporter (500) - 50 UN 0 SN
                          0.533  (± 0.0%) i/s -      6.000  in  11.253824s
PuppetFactImporter (500) - 50 UN 25 SN
                          0.509  (± 0.0%) i/s -      6.000  in  11.799898s
StructuredFactImporter (200) - 0 UN 0 SN
                          1.017  (± 0.0%) i/s -     11.000  in  10.817775s
StructuredFactImporter (200) - 0 UN 25 SN
                          0.950  (± 0.0%) i/s -     10.000  in  10.538463s
StructuredFactImporter (200) - 50 UN 0 SN
                          0.899  (± 0.0%) i/s -      9.000  in  10.010366s
StructuredFactImporter (200) - 50 UN 25 SN
                          0.828  (± 0.0%) i/s -      9.000  in  10.889116s
StructuredFactImporter (500) - 0 UN 0 SN
                          0.552  (± 0.0%) i/s -      6.000  in  10.896482s
StructuredFactImporter (500) - 0 UN 25 SN
                          0.535  (± 0.0%) i/s -      6.000  in  11.225326s
StructuredFactImporter (500) - 50 UN 0 SN
                          0.518  (± 0.0%) i/s -      6.000  in  11.593115s
StructuredFactImporter (500) - 50 UN 25 SN
                          0.493  (± 0.0%) i/s -      5.000  in  10.136886s
Memory stats
Total objects allocated: 257714150
Total heap pages allocated: 6811
```

This patch:

```
PuppetFactImporter (200) - 0 UN 0 SN
                          2.363  (± 0.0%) i/s -     24.000  in  10.388293s
PuppetFactImporter (200) - 0 UN 25 SN
                          1.983  (± 0.0%) i/s -     20.000  in  10.124501s
PuppetFactImporter (200) - 50 UN 0 SN
                          1.989  (± 0.0%) i/s -     20.000  in  10.073955s
PuppetFactImporter (200) - 50 UN 25 SN
                          1.707  (± 0.0%) i/s -     18.000  in  10.564209s
PuppetFactImporter (500) - 0 UN 0 SN
                          0.997  (± 0.0%) i/s -     10.000  in  10.153532s
PuppetFactImporter (500) - 0 UN 25 SN
                          0.939  (± 0.0%) i/s -     10.000  in  10.668852s
PuppetFactImporter (500) - 50 UN 0 SN
                          0.926  (± 0.0%) i/s -     10.000  in  10.813991s
PuppetFactImporter (500) - 50 UN 25 SN
                          0.869  (± 0.0%) i/s -      9.000  in  10.367434s
StructuredFactImporter (200) - 0 UN 0 SN
                          2.447  (± 0.0%) i/s -     25.000  in  10.231096s
StructuredFactImporter (200) - 0 UN 25 SN
                          2.034  (± 0.0%) i/s -     21.000  in  10.335431s
StructuredFactImporter (200) - 50 UN 0 SN
                          1.988  (± 0.0%) i/s -     20.000  in  10.076620s
StructuredFactImporter (200) - 50 UN 25 SN
                          1.701  (± 0.0%) i/s -     17.000  in  10.006774s
StructuredFactImporter (500) - 0 UN 0 SN
                          1.039  (± 0.0%) i/s -     11.000  in  10.589648s
StructuredFactImporter (500) - 0 UN 25 SN
                          0.949  (± 0.0%) i/s -     10.000  in  10.542682s
StructuredFactImporter (500) - 50 UN 0 SN
                          0.933  (± 0.0%) i/s -     10.000  in  10.727191s
StructuredFactImporter (500) - 50 UN 25 SN
                          0.865  (± 0.0%) i/s -      9.000  in  10.411843s
Memory stats
Total objects allocated: 152487435
Total heap pages allocated: 4411
```

lzap commented:
The errors looks unrelated.

dLobatog commented on app/services/structured_fact_importer.rb:
```diff
@@ -24,19 +24,21 @@ def preload_fact_names
   end
 
-  def create_fact_name(fact_names, name, fact_value)
+  def find_or_create_fact_name(name, value)
```
Make `def .... (name, value=nil)` , it's more intuitive  as it just creates the FactName w/o passing nil as you do in line https://github.com/theforeman/foreman/pull/3667/files#diff-509173a705422d0025bd1a6c8a630c72R30

dLobatog reviewed (commented):
@lzap Preliminary :+1: and ready to merge except for that comment on `find_or_create_fact_name` - I want to test on MySQL but I don't expect disparate results. Thanks for this!

lzap commented:
Sure, done. Make sure you test against latest `develop` which includes https://github.com/theforeman/foreman/pull/3852.

dLobatog commented on app/services/structured_fact_importer.rb:
```diff
@@ -24,19 +24,21 @@ def preload_fact_names
     if name.include?(FactName::SEPARATOR)
       parent_name = /(.*)#{FactName::SEPARATOR}/.match(name)[1]
-      parent_fact = create_fact_name(fact_names, parent_name, nil)
+      parent_fact = find_or_create_fact_name(parent_name, nil)
```
No need to pass `nil` here

dLobatog commented:
On MySQL, before this PR:

```
PuppetFactImporter (200) - 0 Unique names 0 Structured names
                          0.879  (± 0.0%) i/s -      9.000  in  10.400460s
PuppetFactImporter (200) - 0 Unique names 25 Structured names
                          0.839  (± 0.0%) i/s -      9.000  in  10.765001s
PuppetFactImporter (200) - 50 Unique names 0 Structured names
                          0.811  (± 0.0%) i/s -      9.000  in  11.105421s
PuppetFactImporter (200) - 50 Unique names 25 Structured names
                          0.750  (± 0.0%) i/s -      8.000  in  10.676619s
PuppetFactImporter (500) - 0 Unique names 0 Structured names
                          0.466  (± 0.0%) i/s -      5.000  in  10.830115s
PuppetFactImporter (500) - 0 Unique names 25 Structured names
                          0.464  (± 0.0%) i/s -      5.000  in  10.804550s
PuppetFactImporter (500) - 50 Unique names 0 Structured names
                          0.446  (± 0.0%) i/s -      5.000  in  11.222318s
PuppetFactImporter (500) - 50 Unique names 25 Structured names
                          0.436  (± 0.0%) i/s -      5.000  in  11.480442s
StructuredFactImporter (200) - 0 Unique names 0 Structured names
                          0.825  (± 0.0%) i/s -      9.000  in  11.280312s
StructuredFactImporter (200) - 0 Unique names 25 Structured names
                          0.512  (± 0.0%) i/s -      5.000  in  10.522463s
StructuredFactImporter (200) - 50 Unique names 0 Structured names
                          0.759  (± 0.0%) i/s -      8.000  in  10.556014s
StructuredFactImporter (200) - 50 Unique names 25 Structured names
                          0.627  (± 0.0%) i/s -      7.000  in  11.331457s
StructuredFactImporter (500) - 0 Unique names 0 Structured names
                          0.399  (± 0.0%) i/s -      5.000  in  12.783046s
StructuredFactImporter (500) - 0 Unique names 25 Structured names
                          0.394  (± 0.0%) i/s -      5.000  in  12.815084s
StructuredFactImporter (500) - 50 Unique names 0 Structured names
                          0.348  (± 0.0%) i/s -      4.000  in  11.622480s
StructuredFactImporter (500) - 50 Unique names 25 Structured names
                          0.411  (± 0.0%) i/s -      5.000  in  12.191670s
Memory stats
Total objects allocated: 212678023
Total heap pages allocated: 4760
```

After this PR

```

PuppetFactImporter (200) - 0 UN 0 SN
                          2.008  (± 0.0%) i/s -     20.000  in  10.208551s
PuppetFactImporter (200) - 0 UN 25 SN
                          1.597  (± 0.0%) i/s -     16.000  in  10.153452s
PuppetFactImporter (200) - 50 UN 0 SN
                          1.685  (± 0.0%) i/s -     17.000  in  10.182653s
PuppetFactImporter (200) - 50 UN 25 SN
                          1.422  (± 0.0%) i/s -     15.000  in  10.638633s
PuppetFactImporter (500) - 0 UN 0 SN
                          0.862  (± 0.0%) i/s -      9.000  in  10.598905s
PuppetFactImporter (500) - 0 UN 25 SN
                          0.831  (± 0.0%) i/s -      9.000  in  10.840475s
PuppetFactImporter (500) - 50 UN 0 SN
                          0.815  (± 0.0%) i/s -      9.000  in  11.053956s
PuppetFactImporter (500) - 50 UN 25 SN
                          0.757  (± 0.0%) i/s -      8.000  in  10.576255s
StructuredFactImporter (200) - 0 UN 0 SN
                          2.126  (± 0.0%) i/s -     22.000  in  10.432616s
StructuredFactImporter (200) - 0 UN 25 SN
                          1.706  (± 0.0%) i/s -     17.000  in  10.077632s
StructuredFactImporter (200) - 50 UN 0 SN
                          1.596  (± 0.0%) i/s -     16.000  in  10.160172s
StructuredFactImporter (200) - 50 UN 25 SN
                          1.481  (± 0.0%) i/s -     15.000  in  10.147651s
StructuredFactImporter (500) - 0 UN 0 SN
                          0.880  (± 0.0%) i/s -      9.000  in  10.307548s
StructuredFactImporter (500) - 0 UN 25 SN
                          0.830  (± 0.0%) i/s -      9.000  in  10.866419s
StructuredFactImporter (500) - 50 UN 0 SN
                          0.765  (± 0.0%) i/s -      8.000  in  10.549480s
StructuredFactImporter (500) - 50 UN 25 SN
                          0.771  (± 0.0%) i/s -      8.000  in  10.385631s
Memory stats
Total objects allocated: 130639777
Total heap pages allocated: 4760
```

dLobatog commented:
Ready to merge if Jenkins tests pass

domcleal commented on app/services/fact_importer.rb:
```diff
@@ -37,11 +37,16 @@ def initialize(host, facts = {})
-    raise ::Foreman::Exception.new(N_("Import of facts failed for host %s"), @host.name) if @error
+    if @error
+      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
+      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s: #{@error}"), @host.name)
```
Interpolation within an extracted string is incorrect (`#{@error}`), and unnecessary here as you're using WrappedException. Just restore the original, `N_("Import of facts failed for host %s")`.

domcleal commented on config/settings.yaml.example:
```diff
@@ -49,3 +49,5 @@
 #    :enabled: false
 #  :templates:
 #    :enabled: true
+#  :audit:
```
This list's alphabetical, it's out of order.

domcleal commented on config/application.rb:
```diff
@@ -176,7 +176,8 @@ class Application < Rails::Application
       :sql => {:enabled => false},
-      :templates => {:enabled => true}
+      :templates => {:enabled => true},
+      :audit => {:enabled => true}
```
This list's alphabetical, it's out of order.

(I'm also unsure why this is called "audit" when the fact import instrumentation isn't related to audits, seems like a poor name.)

lzap commented:
Fixed your comments, thank.

The idea behind audit is to have a logger which can server as second level audit. We have a table in our database, but that one does not fit for arbitrary audit logs, it is only suitable for entries created by the gem. I would like to start adding more audit entries for example the most important - contents of rendered template. There are other candidates as well.

dLobatog commented:
Merged as 46e1ea9f704f41afbd0373e4b8fc54313ac6b397, thanks @lzap!
